### DISCUSSION TITLE
Increase image brightness when one of the bands is zero

### DISCUSSION THREAD

**Discussion by idanya** [libvips/libvips#2530] [General]
Hello, 

I am trying to do some manipulations of an image, part of them is brightness changes. 
For that, I'm using `Linear` with a multiplier.
I've notices an issue when trying to increase brightness of images that has (at least) one of the bands set to zero. 

Example image 0 RGB(0, 255, 255):
![twitter](https://user-images.githubusercontent.com/10362893/141344907-6cb9085e-822c-4239-b4d6-faa99893298c.png)

Decreasing brightness works as expected. Command: 
```
vips linear twitter.png twitter.out.0.4.png 0.4 0
```
Output:
![twitter out 0 4](https://user-images.githubusercontent.com/10362893/141345071-b3d64890-24cb-4119-a3b6-41ec91b614f4.png)

But trying to increase the brightness with the command: 
```
vips linear twitter.png twitter.out.1.7.png 1.7 0
```
Gives the output:
![twitter out 1 7](https://user-images.githubusercontent.com/10362893/141345166-e201b50d-a13d-429e-8c87-d4070394c77a.png)

Which is the same as the source. I assume that is because the R channel is 0 and multiplying it by any number will give 0, and G and B are both at their max, so output cannot change.

But, doing the same with Photoshop:
![Screen Shot 2021-11-11 at 19 52 00](https://user-images.githubusercontent.com/10362893/141345645-8146d1e5-813d-40f6-afe1-5003294ef306.png)

Gives this RGB(178,255,255):
![twitter photoshop 1 7](https://user-images.githubusercontent.com/10362893/141345578-b1849d7b-d07a-4bbf-94f9-69a02f0b8c98.png)

Is there a way to get to the same result with libvips? 
I've tested photoshop with some values and it seems like it set the minimum value for the R band to something between 2 and 3 although it is actually 0. setting a random MIN doesn't feel right. Is there another method for brightness change that is not a multiplication of all bands?

Note: converting colorspace to HSV/Lch and multiplying it gives the same result.

**Comment by jcupitt** [comment]
Hi @idanya,

Adjusting L of LCh seems to do something like what photoshop is doing. For example:

```
$ vips colourspace twitter.png t1.v lch
$ vips linear t1.v t2.v "1.4 1 1 1" "0 0 0 0"
$ vips colourspace t2.v x.png srgb
```

To make:

![x](https://user-images.githubusercontent.com/580843/141384704-442349c7-966f-4b41-81d5-0a3b7b4686d3.png)

**Reply by idanya** [reply]
Hi @jcupitt,

In your example, you use Lch and increase by 1.4 to get the same result as 1.7 in HSV. 
How did you get to 1.4? just trying out numbers or any calculation?

Lch is not possible for me as I also need to do some Hue manipulations that will have to have very different values between the 2 colorspaces.

**Reply by jcupitt** [reply]
I just guessed a number at random.

Perhaps you could adjust lightness in LCh (or Lab) and hue in HSV?

**Reply by idanya** [reply]
The first thing that came to mind :)
This solution could work if I know how to translate the lightness value (HSV 1.7 -> LCH 1.4).
With your test you got to RGB(179,255,255) while photoshop got to RGB(178,255,255) - that was a good hit, that's why I thought you have some kind of formula.

**Reply by jcupitt** [reply]
It won't be straightforward --- L is a power function of G.

**Reply by idanya** [reply]
Interesting. 
Photoshop and GIMP do it pretty much the same. it can't be `linear` nor `HSV` but somehow they are aligned. 
Is there another "agreed-up-on" way to do these manipulations?

**Reply by jcupitt** [reply]
I guess the gimp people spent some time looking at photoshop behaviour and reimplemented it. I don't think there's a standard, except adjusting L in LAB or LCh.
